app/code/Magento/CatalogInventory/Model/ResourceModel/Indexer/Stock/DefaultStock.php

```diff
@@ -305,14 +305,16 @@ protected function _updateIndex($entityIds)
                 'stock_status' => (int)$row['status'],
             ];
             if ($i % 1000 == 0) {
+                $this->deleteOldRecords(array_column($data, 'product_id'));
                 $this->_updateIndexTable($data);
                 $data = [];
             }
         }
 
-        $this->deleteOldRecords($entityIds);
-        $this->_updateIndexTable($data);
-
+        if (is_array($data) && !empty($data)) {
```

Comment:
If we are considering fix for the following issue https://github.com/magento/magento2/issues/12205, it's already fixed withing a scope of 2.2-develop. It should be enough to remove all records at the very beginning of the index update operation without performing a separate delete in a loop. 

Please, check the latest commit here https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/CatalogInventory/Model/ResourceModel/Indexer/Stock/DefaultStock.php for the clarification.

Thank you